### ISSUE TITLE
Better razor panel init behavior for procedually creating panels

### ISSUE THREAD

**Issue by DoctorGurke** [sboxgame/issues#2579]
### For?

S&Box

### What can't you do?

Previously, we were able to use a panel class' constructor to init or assign template properties when procedually creating panels via code. This is no longer the case with razor panel classes, as properties won't be initialized properly, instead we have to use the `OnAfterTreeRender(bool firsttime)` override (what a mouth-full).

Previously:

```csharp
public ExamplePanel( string name, string image )
{
    Name.SetText( name );
    Image.SetTexture( image );
}
```
Now:
```csharp
private string name;
private string image;

public ExamplePanel( string name, string image )
{
    this.name = name;
    this.image = image;
}

protected override void OnAfterTreeRender( bool firstTime )
{
    base.OnAfterTreeRender( firstTime );

    if( !firstTime ) 
        return;

    Name.SetText( name );
    Image.SetTexture( image );
}
```
This is kinda smelly and makes my life a lot more annoying.

### How would you like it to work?

I have no clue how this works in the backend and I am sure there's technical implications for this kind of init behavior, but as it stands this is kind of terrible. A separate override for first time init would already be a lot cleaner, but this behavior is just kind of bad for procedually creating panels from code in general.

### What have you tried?

N/A

**Comment by Fortune117** [comment]
I've run into this issue myself, it means you need lots of extra boiler plate code and is definitely my biggest gripe with razor so far.

**Comment by matekdev** [comment]
Yes, I just encountered this myself when trying to swap over my UI to razor. It would be nice if we could just init the properties in the constructor.

**Comment by DoctorGurke** [comment]
This is especially relevant now that templates are obsolete and we're being forced into this behavior. Can we get some official response for this please?

**Comment by DoctorGurke** [comment]
This problem applies to procedually setting/updating and initializing panels from code.
A more complex example, where I set a panel's content based on some resource through code:
```csharp
private bool _initialized;
private MyResource _cachedResource;

protected override void OnAfterTreeRender( bool firstTime )
{
    if(firstTime)
    {
        _initialized = true;

        if(_cachedResource is not null)
            SetContent(_cachedResource);

        _cachedResource = null;
    }
}

public void SetContent(MyResource resource)
{
    if(!_initialized)
        _cachedResource = resource;
    
    // assign @refs based on resource etc 
}
```

**Comment by DoctorGurke** [comment]
for the record, this is what I need to do now, previously I could just set the content from the constructor or hotload directly.

**Comment by garrynewman** [comment]
You're gonna have to be more explicit about what you're doing because I don't understand the problem at all

**Comment by DoctorGurke** [comment]
```csharp
var panel = container.AddChild<ResourcePanel>(resource);
```
where the constructor would previously call SetContent with the resource

**Comment by garrynewman** [comment]
but if you want to create the UI in code why are you trying to make it use razor?

**Comment by DoctorGurke** [comment]
I'm assigning things such as label text etc, for panels that have a layout made in razor. I don't want to make the entire layout of a complex panel in code if I can use html for it.

**Comment by garrynewman** [comment]
But why are you trying to make it half in razor and half in code?

**Comment by DoctorGurke** [comment]
I want to be able to layout the fundamental elements in html because doing that for anything with more than a couple elements in code sucks and that's what html is for.
I also want some procedual behavior for updating and creating these panels at runtime.

**Comment by garrynewman** [comment]
Can you give me a simple example of what you're doing this for. If there's something that you can't do in Razor I'm all ears, I can't imagine a scenario where it would be preferable to skip out and create a bunch of sub-elements in code.

**Comment by DoctorGurke** [comment]
I'm not creating sub elements in code, I'm creating things such as list entries in code.
example, Item Entries in a shop UI

**Comment by garrynewman** [comment]
Sure but in that case why aren't you doing it via foreach in razor, since it makes that stuff so much easier?

**Comment by matekdev** [comment]
For what it's worth, I realized that I was approaching `razor` as if it was HTML templates (and I'm assuming everyone else who liked the post). At least relating to the original snippet in this issue, I realized how you'd approach it via using `razor`.

**Comment by cr4yz** [comment]
> For what it's worth, I realized that I was approaching `razor` as if it was HTML templates (and I'm assuming everyone else who liked the post). At least relating to the original snippet in this issue, I realized how you'd approach it via using `razor`.

Yep I think that is a big part of the confusion.  The razor workflow is similar but not seamless, and approaching things the razor way might not be immediately obvious

**Comment by DoctorGurke** [comment]
I'd probably use razor embedded code more if the intellisense was nicer to use, but I'll see if converting to that solves some of these issues. 
I still think for procedually initializing panels (without @ref being a concern) this behavior is extremely bad. I don't want to make a backing field for every property just because I can't use the constructor anymore. Something like base Chat and ChatEntry would have the same problem if they used razor. And I really want to use html layout for anything more complex than an image and 2 labels.

To elaborate on that example, adding entries from a concmd won't work if I'm trying to initialize panel properties that are @ref. Basically what a chatbox with chatentries would want to do.

**Comment by DoctorGurke** [comment]
Also scratch that, intellisense outside of razor in partial classes also seems to utterly break down now, not sure if it was always this bad, but it's quite awful.

**Comment by DoctorGurke** [comment]
Test example for what I mean:

```csharp
[ConCmd.Client]
public static void AddEntryTest( string message )
{
	if ( Current?.ContainerPanel is null )
		return;

	var entry = Current.ContainerPanel.AddChild<Entry>();

	entry.Message.Text = message; // NRW because the Message @ref won't be initialized
}
```

This seems to indicate that it's not initializing these panels immediately?

**Comment by cr4yz** [comment]
References are not set before the constructor and unless you've created them, they won't exist.  That is expected behaviour.

Here are two different methods to do a chatbox in razor: [example1](https://github.com/Facepunch/sbox-unicycle-frenzy/blob/fb7f3dcb47a7ca7b97fd2dd2cfedc043393cdbf4/code/UI/Chatbox/UfChat.razor#L43) [example2](https://github.com/Facepunch/sbox-strafe/blob/d92780fc2914ea1a0adf6770217c20b2334f3084/code/UI/Chat/Chatbox.razor#L49)

**Comment by DoctorGurke** [comment]
The latest example tries to assign the value after the constructor.

**Comment by cr4yz** [comment]
It stores the name & message in a collection to be displayed with a foreach loop, [see here](https://github.com/Facepunch/sbox-strafe/blob/d92780fc2914ea1a0adf6770217c20b2334f3084/code/UI/Chat/Chatbox.razor#L7)

**Comment by DoctorGurke** [comment]
Very nice, I'm pointing out the example I proved just a few messages above however. Trying to access a @ref *after* the constructor.

![sbox-dev_NVBvrOxuJ8](https://user-images.githubusercontent.com/12602100/206553740-5ca6921c-5ec1-4598-b09a-1b493779d6d8.png)


Full example code:

```csharp
public partial class EntryContainer
{
	[ConCmd.Client]
	public static void AddEntryTest( string message )
	{
		if ( Current is null )
			return;

		var entry = new TestEntry();
		Current.AddChild( entry );

		if ( entry.Message is null )
			Log.Error( $"@ref not initialized, even after the constructor" );

		entry.Message.Text = message;
	}
}
```

```
@using Sandbox;
@using Sandbox.UI;

<root/>

@code {
	private static EntryContainer Current;

	public EntryContainer()
	{
		Current = this;
	}
}
```

```
@using Sandbox;
@using Sandbox.UI;

<root>
	<label @ref="Message"/>
</root>

@code {
	public Label Message { get; set; }
}
```

**Comment by DoctorGurke** [comment]
that example can be trimmed down a ton for this exact issue, just happened to be the test case I had in my project

**Comment by Remscar** [comment]
I think you're better off creating a list of strings (or some data class), having that as a member of your `EntryContainer` class, and then doing the razor `foreach` over that list of strings, and then creating a `TestEntry` element for each string in that list and setting the necessary parameters in razor.

I ran into a similar situation, here is the before/after code:

**Before:**

`WrenchInterface.cs`
```cs
public partial class WrenchInterface : Panel
{

    public Panel PropListPanel;

    protected override void OnAfterTreeRender(bool firstTime)
    {
        if (!firstTime)
            return;

        PopulateBuildList();
    }

    public void PopulateBuildList()
    {
        var props = BuildableProp.Props;

        for (var i = 0; i < props.Count; ++i)
        {
            var prop = props[i];
            var panel = PropListPanel.AddChild<Prop>();
            panel.PropData = prop;
            panel.PropIndex = i;
        }
    }
}
```

`WrenchInterface.razor`
```html
<div class="prop-list" @ref="PropListPanel">
```

Where `BuildableProp` is:
```cs
public class BuildableProp
{
    public string Name { get; set; }
    public int Cost { get; set; }
    public string ModelPath { get; set; }

    public static List<BuildableProp> Props {get; set;}
}
```

**After:**

No `WrenchInterface.cs` code

`WrenchInterface.razor`
```cs
<div class="prop-list">
    @foreach ( var (item, index) in BuildableProp.Props.WithIndex() ) {
        <Prop PropData="@item" PropIndex="@index" />
    }
</div>
```

**Comment by DoctorGurke** [comment]
After all that back and forth, I'm sure most of my issues are solved just by using razor in the ways demonstrated. But surely it's not normal that panel references aren't initialized/linked even after the constructor is ran? How do panels get initialized right now? Are they only valid on the next frame/tick/done in async? It's really weird behavior and I'd expect something as simple as that to work.

**Comment by garrynewman** [comment]
Yeah it's normal. The Razor panel isn't "built" until its Tick is ran. The refs get set during build.

**Comment by DoctorGurke** [comment]
That just seems wrong to me... Tell me I'm crazy for thinking that's odd. Considering that I'm facing a rock solid opposition here, I'd like to assume this is some sort of technical limitations with how razor panels are created? 
All in all this makes me even sadder that html templates are being completely cut out while the alternative is not only extensively more involved and complex, but also doesn't even fully work yet with the usual IDEs. Mostly referring to the absolutely abysmal intellisense and hot load perf...

But fine, I'll yield. I tried my best to make a point for simple, fast and flexible html Templates already, which didn't work out either.
Thank you everybody for your  consideration and inputs. Closing the issue.
